File: app/src/main/java/com/stho/software/nyota/universe/TLE.java
package com.stho.software.nyota.universe;

import com.stho.software.nyota.utilities.Formatter;
import com.stho.software.nyota.utilities.JulianDay;
import com.stho.software.nyota.utilities.Radian;
import com.stho.software.nyota.utilities.UTC;

import java.util.Date;

/**
 * Two line elements
 */
public class TLE {
    private String elements;

    /** Epoch of the TLE as Julian Day */
    public double Epoch;

    public double bstar;

    /** Inclination (in radian) */
    public double xincl;

    /** Right Ascension of the Ascending Node (in radian) */
    public double xnodeo;

    /** Eccentricity (decimal) */
    public double eo;

    /** Argument of Perigee (in radian) */
    public double omegao;

    /** Mean Anomaly (in radian) */
    public double xmo;

    /** Mean Motion (in radian per minute) */
    public double xno;

    public int SatelliteNumber;

    public int RevolutionNumber;

    public String InternationalDesignator;

    /** Mean Distance in km */
    public double MeanDistanceFromEarth;

    /** Mean Motion (in revolutions per day) */
    public double RevolutionsPerDay;


    /** If the TLE is still valid or if aa new download is required */
    public boolean isOutdated()
    {
        UTC now = UTC.forNow();
        return ((now.getJulianDay() - this.Epoch) > 3);
    }

    public String serialize()
    {
        return this.elements;
    }

    public boolean deserialize(String elements) {
        if (!TLEParser.tryParse(this, elements))
            return false;

        this.elements = elements;
        return true;
    }

    public Date getDate() {
        return JulianDay.toUTC(Epoch).getTime();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        sb.append(String.format("NORAD = %d\n", this.SatelliteNumber));
        sb.append(String.format("Epoch (UTC) = %s\n", Formatter.dateTime.format(JulianDay.toUTC(this.Epoch))));
        sb.append(String.format("Eccentricity = %$.3f", this.eo));
        sb.append(String.format("Inclination = %$.1f°", Radian.toDegrees(this.xincl)));
        sb.append(String.format("Mean Height = %$.f km", this.MeanDistanceFromEarth));

        double difference = this.eo * ( (this.MeanDistanceFromEarth + Algorithms.EARTH_RADIUS));

        sb.append(String.format("Perigee Height = %1$.f km", this.MeanDistanceFromEarth - difference));
        sb.append(String.format("Apogee Height = %$.f km", this.MeanDistanceFromEarth + difference));
        sb.append(String.format("Right Ascension of Ascend Node = %$.1f°", Radian.toDegrees(this.xnodeo)));
        sb.append(String.format("Argument of Perigee = %$.1f°", Radian.toDegrees(this.omegao)));
        sb.append(String.format("Revs Per Day = %$.1f", this.RevolutionsPerDay));
        sb.append(String.format("Mean anomaly = %$.1f°", Radian.toDegrees(this.xmo)));

        return sb.toString();
    }
}

File: app/src/main/java/com/stho/nyota/sky/universe/TLE.kt
package com.stho.nyota.sky.universe

import com.stho.nyota.sky.utilities.Formatter
import com.stho.nyota.sky.utilities.JulianDay.toUTC
import com.stho.nyota.sky.utilities.Radian.toDegrees
import com.stho.nyota.sky.utilities.UTC
import com.stho.nyota.sky.utilities.UTC.Companion.forNow
import java.security.InvalidParameterException
import java.util.*

/**
 * Two line elements
 */
data class TLE(
        val elements: String,
        val Epoch: Double, /* Epoch of the TLE as Julian Day  */
        val bstar: Double,
        val xincl: Double,  /* Inclination (in radian) */
        val xnodeo: Double, /* Right Ascension of the Ascending Node (in radian) */
        val eo: Double, /* Eccentricity (decimal)  */
        val omegao: Double, /* Argument of Perigee (in radian)  */
        val xmo: Double, /* Mean Anomaly (in radian)  */
        val xno: Double, /* Mean Motion (in radian per minute)  */
        val noradSatelliteNumber: Int, /* NORAD satellite number */
        val revolutionNumber: Int,
        val internationalDesignator: String,
        val meanDistanceFromEarth: Double, /* Mean Distance in km  */
        val revolutionsPerDay: Double) { /* Mean Motion (in revolutions per day) */

    /** If the TLE is still valid or if aa new download is required  */
    val isOutdated: Boolean
        get() = isOutdated(this)

    fun serialize(): String = elements

    val date: Date
        get() = toUTC(Epoch).time

    override fun toString(): String {
        val difference = eo * (meanDistanceFromEarth + Algorithms.EARTH_RADIUS)
        val sb = StringBuilder()
        sb.append(String.format("NORAD = %d", noradSatelliteNumber))
        sb.append(EOL)
        sb.append(String.format("Epoch (UTC) = %s", Formatter.formatDateTime.format(toUTC(Epoch))))
        sb.append(EOL)
        sb.append(String.format("Eccentricity = %.3f", eo))
        sb.append(EOL)
        sb.append(String.format("Inclination = %.1f°", toDegrees(xincl)))
        sb.append(EOL)
        sb.append(String.format("Mean Height = %.f km", meanDistanceFromEarth))
        sb.append(EOL)
        sb.append(String.format("Perigee Height = %.f km", meanDistanceFromEarth - difference))
        sb.append(EOL)
        sb.append(String.format("Apogee Height = %.f km", meanDistanceFromEarth + difference))
        sb.append(EOL)
        sb.append(String.format("Right Ascension of Ascend Node = %.1f°", toDegrees(xnodeo)))
        sb.append(EOL)
        sb.append(String.format("Argument of Perigee = %.1f°", toDegrees(omegao)))
        sb.append(EOL)
        sb.append(String.format("Revolutions per Day = %.1f", revolutionsPerDay))
        sb.append(EOL)
        sb.append(String.format("Mean anomaly = %.1f°", toDegrees(xmo)))
        return sb.toString()
    }

    fun toSummaryString(): String {
        val difference = eo * (meanDistanceFromEarth + Algorithms.EARTH_RADIUS)
        val sb = StringBuilder()
        sb.append(String.format("Height %.0f - %.0f km", (meanDistanceFromEarth - difference), (meanDistanceFromEarth + difference)))
        sb.append(EOL)
        sb.append(String.format("Revolutions per Day = %.1f", revolutionsPerDay))
        return sb.toString()
    }

    companion object {

        private const val DAYS_TLE_KEEP_VALID: Int = 3
        private const val EOL: String = "\n"

        private fun isOutdated(tle: TLE): Boolean {
            val now: UTC = forNow()
            return (now.julianDay - tle.Epoch) > DAYS_TLE_KEEP_VALID
        }

        /*
            Deserializes the 2-line-elements into a TLE structure or throws an InvalidParameterException if parsing the elements fails
         */
        fun deserialize(elements: String): TLE {
            val tle = TLEParser().parseTLE(elements)
            if (tle != null) {
                return tle
            } else {
                throw InvalidParameterException("Invalid TLE: $elements")
            }
        }
    }
}